internal/network/connectivity_groups.go

```diff
@@ -142,116 +143,235 @@ func (g groupId) isLess(other groupId) bool {
+type connectivityGroupList struct {
+	groupsBySize map[int][]*connectivityGroup
+}
 
-		// Add the sets in the pending list to the groupList
-		for _, set := range pendingSets {
-			groupList.addSet(set)
-		}
+func newConnectivityGroupList() *connectivityGroupList {
+	return &connectivityGroupList{
+		groupsBySize: make(map[int][]*connectivityGroup),
 	}
+}
 
-	// Second iteration - Per groupList element. count the number of candidates that are part of that element.
-	// Since every candidate represents a host, if the number of participants == the set size, then there is a full
-	// mesh connectivity between the members
-	for _, candidate := range groupCandidates {
-		for _, cs := range groupList.groups {
-			if candidate.set.isSupersetOf(cs.set) && cs.set.containsElement(candidate.me) {
-				cs.count++
-			}
+func (c *connectivityGroupList) findGroup(cg *connectivityGroup) *connectivityGroup {
+	groupsForSize, exists := c.groupsBySize[cg.len()]
+	if !exists {
+		return nil
+	}
+	for _, g := range groupsForSize {
+		if g.equivalent(cg) {
+			return g
 		}
 	}
-	return groupList
+	return nil
 }
 
-func filterFullMeshGroups(groupList connectivityGroupList) []*connectivitySet {
-	ret := make([]*connectivitySet, 0)
-	for _, r := range groupList.groups {
-		// Add only sets with full mesh connectivity
-		if r.set.len() == r.count {
-			ret = append(ret, r.set)
-		}
+func (c *connectivityGroupList) addGroup(cg *connectivityGroup) {
+	size := cg.len()
+	c.groupsBySize[size] = append(c.groupsBySize[size], cg)
+}
+
+func (m *majorityGroupCalculator) toConnectivityGroup(candidate *groupCandidate) *connectivityGroup {
+	ret := &connectivityGroup{
+		set:                candidate.set,
+		participatingHosts: NewConnectivitySet(m.numHosts),
 	}
+	_ = ret.participatingHosts.add(candidate.me)
 	return ret
 }
 
-// Create sorted list of connectivity sets.  The sort is by set size (descending)
-func createConnectivityGroups(groupCandidates []groupCandidate) []*connectivitySet {
-	groupList := createGroupList(groupCandidates)
-	ret := filterFullMeshGroups(groupList)
+func (c *connectivityGroupList) addOrMergeGroup(cg *connectivityGroup) {
+	foundGroup := c.findGroup(cg)
+	if foundGroup != nil {
+		foundGroup.mergeParticipants(cg)
+	} else {
+		c.addGroup(cg)
+	}
+}
 
-	// Sort by set size descending, which means the largest group first.
+func (c *connectivityGroupList) largestGroup() *connectivityGroup {
+	size := c.biggestGroupSize()
+	if size < 3 {
+		return nil
+	}
+	groupsBySize := c.groupsBySize[size]
+	var ret []*connectivityGroup
+	var numParticipatingHosts int
+	for _, cg := range groupsBySize {
+
+		// Take only the participants with the highest number of participating hosts
+		switch {
+		case cg.numParticipatingHosts() > numParticipatingHosts:
+			numParticipatingHosts = cg.numParticipatingHosts()
+			ret = nil
+			fallthrough
+		case cg.numParticipatingHosts() == numParticipatingHosts:
+			ret = append(ret, cg)
```

Comment:
Instead of appending and later sorting by id, consider updating ret if
`if cg.id.isLess(currentId)...`

Reply:
OK